account_invoice_merge/models/account_invoice.py

```diff
@@ -0,0 +1,240 @@
+
+                o_line = invoice_infos["invoice_line_ids"].setdefault(line_key, {})
+
+                if o_line:
+                    # merge the line with an existing line
+                    o_line["quantity"] += invoice_line.quantity
+                else:
+                    # append a new "standalone" line
+                    o_line["quantity"] = invoice_line.quantity
+
+        allinvoices = []
+        allnewinvoices = []
+        invoices_info = {}
+        old_invoices = self.env["account.move"]
+        qty_prec = self.env["decimal.precision"].precision_get(
+            "Product Unit of Measure"
+        )
+        for _invoice_key, (invoice_data, old_ids) in new_invoices.items():
+            # skip merges with only one invoice
+            if len(old_ids) < 2:
+                allinvoices += old_ids or []
+                continue
+            # cleanup invoice line data
+            for key, value in invoice_data["invoice_line_ids"].items():
+                value.update(dict(key))
+
+            if remove_empty_invoice_lines:
+                invoice_data["invoice_line_ids"] = [
+                    (0, 0, value)
+                    for value in invoice_data["invoice_line_ids"].values()
+                    if not float_is_zero(value["quantity"], precision_digits=qty_prec)
+                ]
+            else:
+                invoice_data["invoice_line_ids"] = [
+                    (0, 0, value) for value in invoice_data["invoice_line_ids"].values()
+                ]
+
+            if date_invoice:
+                invoice_data["invoice_date"] = date_invoice
+
+            # create the new invoice
+            newinvoice = self.with_context(is_merge=True).create(invoice_data)
+            invoices_info.update({newinvoice.id: old_ids})
+            allinvoices.append(newinvoice.id)
+            allnewinvoices.append(newinvoice)
+            # cancel old invoices
+            old_invoices = self.env["account.move"].browse(old_ids)
+            old_invoices.with_context(is_merge=True).button_cancel()
+
+        # Make link between original sale order
+        # None if sale is not installed
+        invoice_line_obj = self.env["account.move.line"]
+        for new_invoice_id in invoices_info:
+            if "sale.order" in self.env.registry:
+                sale_todos = old_invoices.mapped(
+                    "invoice_line_ids.sale_line_ids.order_id"
+                )
+                for org_so in sale_todos:
+                    for so_line in org_so.order_line:
+                        invoice_line = invoice_line_obj.search(
+                            [
+                                ("id", "in", so_line.invoice_lines.ids),
+                                ("move_id", "=", new_invoice_id),
+                            ]
+                        )
+                        if invoice_line:
+                            so_line.write({"invoice_lines": [(6, 0, invoice_line.ids)]})
```

Comment:
Should better be so_line.write({"invoice_lines": [(4, ilid) for ilid in invoice_line.ids]})
So that sale order lines doesn't loose reference with old invoices in case of multiple invoices.

Reply:
@Martronic-SA I updated this with your suggestion